### PR TITLE
fix: toBignumber conversion error with high balance

### PR DESCRIPTION
**Pull request by siibars**
## **Description**
There is a bignumber conversion attempt which is failing, this change fixes this.

## **Related issues**

Fixes: STAKE-848

## **Manual testing steps**

1. connect with an account with a high ETH balance 
2. enable the native staking feature flag
3. click stake to get to the StakeinputView
 

### **Before**

image

View: Root
Error: [number-to-bn] while converting number "3.487400189415765903484e+21" to BN.js instance, error: invalid number value. Value must be an integer, hex string, BN or BigNumber instance. Note, decimals are not supported.

### **After**

This issue doesn't happen. 

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by Matt561 — DISMISSED** [review]

File: app/components/UI/Stake/hooks/useBalance.ts [hunk 1]
```diff
@@ -50,7 +51,10 @@ const useBalance = () => {
   );
 
   const { pooledStakesData } = usePooledStakes();
-  const assets = pooledStakesData.assets ?? 0;
+  const assets = useMemo(
+    () => new BN(pooledStakesData.assets),
```
Comment (legobeat) [inline app/components/UI/Stake/hooks/useBalance.ts:55]: It would be better if we can use `hexToBN` from `@metamask/controller-utils` if possible

**Review by amitabh94 — DISMISSED** [review]

File: app/components/UI/Stake/hooks/useBalance.ts (hunk 1 above)
Comment (siibars) [reply app/components/UI/Stake/hooks/useBalance.ts]: yes done

File: app/components/UI/Stake/hooks/useBalance.ts [hunk 2]
```diff
@@ -50,8 +50,10 @@ const useBalance = () => {
   );
 
   const { pooledStakesData } = usePooledStakes();
-  const assets = pooledStakesData.assets ?? 0;
-
+  const assets = useMemo(
+    () => hexToBN(pooledStakesData.assets).toString('hex'),
+    [pooledStakesData],
+  );
```
Comment (Daniel-Cross) [inline app/components/UI/Stake/hooks/useBalance.ts:56]: Is the `useMemo` necessary? Does it meet the criteria for allocating system memory?

https://www.notion.so/metamask-consensys/React-Native-Contributor-Guidelines-127f86d67d6880adb3f4e9ec784c0d36

**Review by Daniel-Cross — CHANGES_REQUESTED** [review]
Check whether the `useMemo` is actually needed. This kind of hook allocates system memory to watch the dependancies.

File: app/components/UI/Stake/hooks/useBalance.ts (hunk 2 above)
Comment (siibars) [reply app/components/UI/Stake/hooks/useBalance.ts:56]: We had a lint error ( react-hooks/exhaustive-deps) when not in a UseMemo [here](https://github.com/MetaMask/metamask-mobile/actions/runs/11504536996/job/32024236985#step:5:9)

**Review by Daniel-Cross — DISMISSED** [review]
Looks good, once the tests are passing. Thanks

**Comment by legobeat** [comment]
> Do you have a link (or would you be able to describe in PR description) the issue this fixes?
> 
> Considering if this is related at all to solving it:
> 
>     * [fix(deps): unpin ethereumjs-abi@0.6,x ethereumjs-util@6.x #11972](https://github.com/MetaMask/metamask-mobile/pull/11972)
> 

@siibars Thanks for updating, much clearer! :+1: 

Looks like it is indeed related. `bn.js` v4 does accept numbers with scientific notation as input to the constructor while v5 throws an error. It seems likely that this was once working correctly when implemented towards `bn.js@4` and then at some point it started failing as the library version used in that particular place got substituted with v5. Unfortunately, the way `BN` is imported from legacy `ethereumjs-util` makes these discrepancies currently not get caught in type checks.


First steps in sorting out this larger mess:
- #11972
- https://github.com/MetaMask/core/pull/4844

**Comment by legobeat** [comment]
@siibars I am trying to track down exactly where the `stakedBalanceWei` (`assets`) string gets constructed - do you have a reference handy?

**Comment by amitabh94** [comment]
> @siibars I am trying to track down exactly where the `stakedBalanceWei` (`assets`) string gets constructed - do you have a reference handy? Is it just being piped unmodified from a server-side API?
> 
> https://github.com/MetaMask/metamask-mobile/actions/runs/11526358054/job/32090342121?pr=12010#step:5:31
> 
> ```
> FAIL app/components/UI/Stake/hooks/useBalance.test.tsx
>   ● useBalance › returns balance and fiat values based on account and pooled stake data
> 
>     expect(received).toBe(expected) // Object.is equality
> 
>     Expected: "5791332670714232000"
>     Received: "05791332670714232000"
> 
>       76 |     expect(result.current.balanceFiat).toBe('$39506172511.60'); // Fiat balance
>       77 |     expect(result.current.balanceFiatNumber).toBe(39506172511.6); // Fiat number balance
>     > 78 |     expect(result.current.stakedBalanceWei).toBe('5791332670714232000'); // No staked assets
>          |                                             ^
>       79 |     expect(result.current.formattedStakedBalanceETH).toBe('5.79133 ETH'); // Formatted ETH balance
>       80 |     expect(result.current.stakedBalanceFiatNumber).toBe(18532.26454); // Staked balance in fiat number
>       81 |     expect(result.current.formattedStakedBalanceFiat).toBe('$18532.26'); //
> 
>       at Object.toBe (app/components/UI/Stake/hooks/useBalance.test.tsx:78:45)
>       at asyncGeneratorStep (node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:17)
>       at _next (node_modules/@babel/runtime/helpers/asyncToGenerator.js:17:9)
>       at node_modules/@babel/runtime/helpers/asyncToGenerator.js:22:7
>       at Object.<anonymous> (node_modules/@babel/runtime/helpers/asyncToGenerator.js:14:12)
> ```

We get this as wei string from the server side API. 
https://github.com/consensys-vertical-apps/va-mmcx-staking-api/blob/81718575f23d03763d8abb991a2379cf54d25041/src/metaMaskPooledStaking/metaMaskPooledStaking.controller.ts#L77

**Comment by amitabh94** [comment]
using new BN(assets) instead of hexToBn(assets) fixes the issue.

**Comment by siibars** [comment]
hi @legobeat 
to give you more details on where stakedBalanceWei comes from :
a. in the mobile it's populated [here](https://github.com/MetaMask/metamask-mobile/blob/2ca359184512dfe999443b3f6b4b01ccbfa963a3/app/components/UI/Stake/hooks/usePooledStakes.ts#L48) with 'pooledStakeData' which comes from an api server call,
b. this call is implemented in the stake-sdk [here](https://github.com/MetaMask/stake-sdk/blob/7c1c655d4fcdb958dc1aa5d432c9dbfa5d5fc19c/src/api/PooledStaking/StakingApiService.ts#L39)
c. The api server implementation makes call to a smart contract method 'getShares' [here ](https://github.com/consensys-vertical-apps/va-mmcx-staking-api/blob/af9f57c7322f1cd6ae7ff64a685171184a025c15/src/metaMaskPooledStaking/metaMaskPooledStaking.service.ts#L238) retrieve an share amount (uint256) and then converts it to a @ethersproject/bignumber to then converts it to a 'bignumber.js' (v9) [here](https://github.com/consensys-vertical-apps/va-mmcx-staking-api/blob/af9f57c7322f1cd6ae7ff64a685171184a025c15/src/metaMaskPooledStaking/metaMaskPooledStaking.service.ts#L189),
d. Finally it multiplies it [here](https://github.com/consensys-vertical-apps/va-mmcx-staking-api/blob/af9f57c7322f1cd6ae7ff64a685171184a025c15/src/metaMaskPooledStaking/metaMaskPooledStaking.service.ts#L190) with conversion rate to get an asset amount 
e. Finally Finally this server api implementation return d as a string [here](https://github.com/consensys-vertical-apps/va-mmcx-staking-api/blob/af9f57c7322f1cd6ae7ff64a685171184a025c15/src/metaMaskPooledStaking/metaMaskPooledStaking.service.ts#L118) using BN.toString 😮

I think this hextobn util [here ](https://github.com/MetaMask/metamask-mobile/blob/8229dbcc07de3506ab3e0d1082502c2618ea1cc9/app/util/number/index.js#L31)make sense but Indeed the BN import should probably be aligned and this file should also be written in ts for more type safety at build time.

@amitabh94 I am curious about d and e working as expected with account with a high share amount, do we have a test for such case ?

cheers

**Comment by amitabh94** [comment]
> hi @legobeat to give you more details on where stakedBalanceWei comes from : a. in the mobile it's populated [here](https://github.com/MetaMask/metamask-mobile/blob/2ca359184512dfe999443b3f6b4b01ccbfa963a3/app/components/UI/Stake/hooks/usePooledStakes.ts#L48) with 'pooledStakeData' which comes from an api server call, b. this call is implemented in the stake-sdk [here](https://github.com/MetaMask/stake-sdk/blob/7c1c655d4fcdb958dc1aa5d432c9dbfa5d5fc19c/src/api/PooledStaking/StakingApiService.ts#L39) c. The api server implementation makes call to a smart contract method 'getShares' [here ](https://github.com/consensys-vertical-apps/va-mmcx-staking-api/blob/af9f57c7322f1cd6ae7ff64a685171184a025c15/src/metaMaskPooledStaking/metaMaskPooledStaking.service.ts#L238) retrieve an share amount (uint256) and then converts it to a @ethersproject/bignumber to then converts it to a 'bignumber.js' (v9) [here](https://github.com/consensys-vertical-apps/va-mmcx-staking-api/blob/af9f57c7322f1cd6ae7ff64a685171184a025c15/src/metaMaskPooledStaking/metaMaskPooledStaking.service.ts#L189), d. Finally it multiplies it [here](https://github.com/consensys-vertical-apps/va-mmcx-staking-api/blob/af9f57c7322f1cd6ae7ff64a685171184a025c15/src/metaMaskPooledStaking/metaMaskPooledStaking.service.ts#L190) with conversion rate to get an asset amount e. Finally Finally this server api implementation return d as a string [here](https://github.com/consensys-vertical-apps/va-mmcx-staking-api/blob/af9f57c7322f1cd6ae7ff64a685171184a025c15/src/metaMaskPooledStaking/metaMaskPooledStaking.service.ts#L118) using BN.toString 😮
> 
> I think this hextobn util [here ](https://github.com/MetaMask/metamask-mobile/blob/8229dbcc07de3506ab3e0d1082502c2618ea1cc9/app/util/number/index.js#L31)make sense but Indeed the BN import should probably be aligned and this file should also be written in ts for more type safety at build time.
> 
> @amitabh94 I am curious about d. and e. working as expected with account with a high share amount / conversion rate with a lot of decimals, do we have tests for such cases ?
> 
> cheers

We will have to check whether if we have specific test cases for decimals but what I know is that we use bignumber.js in Staking api which is used to deal with big numbers with decimal places.

**Review by amitabh94 — APPROVED** [review]

**Pull request merged**